src/main/java/com/datastax/oss/driver/internal/core/loadbalancing/LatencySensitiveLoadBalancingPolicy.java

```diff
@@ -0,0 +1,318 @@
+      this.timestamp = timestamp;
+      this.average = average;
+      this.nbMeasure = nbMeasure;
+    }
+  }
+
+  protected static class NodeLatencyTracker {
+
+    private final long thresholdToAccount;
+    //    private final double localScale;
+    private final AtomicReference<TimestampedAverage> current =
+        new AtomicReference<TimestampedAverage>();
+
+    private final long scale = TimeUnit.MILLISECONDS.toNanos(100);
+
+    NodeLatencyTracker(long thresholdToAccount) {
+      //      this.localScale = (double) localScale; // We keep in double since that's how we'll use
+      // it.
+      this.thresholdToAccount = thresholdToAccount;
+    }
+
+    public void add(long newLatencyNanos) {
+      TimestampedAverage previous, next;
+      do {
+        previous = current.get();
+        next = computeNextAverage(previous, newLatencyNanos);
+      } while (next != null && !current.compareAndSet(previous, next));
+    }
+
+    private TimestampedAverage computeNextAverage(
+        TimestampedAverage previous, long newLatencyNanos) {
+
+      long currentTimestamp = System.nanoTime();
+
+      long nbMeasure = previous == null ? 1 : previous.nbMeasure + 1;
+      if (nbMeasure < thresholdToAccount)
+        return new TimestampedAverage(currentTimestamp, -1L, nbMeasure);
+
+      if (previous == null || previous.average < 0)
+        return new TimestampedAverage(currentTimestamp, newLatencyNanos, nbMeasure);
+
+      // Note: it's possible for the delay to be 0, in which case newLatencyNanos will basically be
+      // discarded. It's fine: nanoTime is precise enough in practice that even if it happens, it
+      // will be very rare, and discarding a latency every once in a while is not the end of the
+      // world.
+      // We do test for negative value, even though in theory that should not happen, because it
+      // seems
+      // that historically there has been bugs here
+      // (https://blogs.oracle.com/dholmes/entry/inside_the_hotspot_vm_clocks)
+      // so while this is almost surely not a problem anymore, there's no reason to break the
+      // computation
+      // if this even happen.
+      long delay = currentTimestamp - previous.timestamp;
+      if (delay <= 0) return null;
+
+      double scaledDelay = ((double) delay) / scale;
+      // Note: We don't use log1p because we it's quite a bit slower and we don't care about the
+      // precision (and since we
+      // refuse ridiculously big scales, scaledDelay can't be so low that scaledDelay+1 == 1.0 (due
+      // to rounding)).
+      double prevWeight = Math.log(scaledDelay + 1) / scaledDelay;
+      long newAverage =
+          (long) ((1.0 - prevWeight) * newLatencyNanos + prevWeight * previous.average);
+
+      return new TimestampedAverage(currentTimestamp, newAverage, nbMeasure);
+    }
+
+    public TimestampedAverage getCurrentAverage() {
+      return current.get();
+    }
+  }
```

Comment:
Not essential by any means but I believe you could move the implementations of the onNodeSuccess and onNodeError methods as well as the classes TimestampedAverage and NodeLatencyTracker into a new class which subclasses DefaultLoadBalancingPolicy.  You could then subclass _that_ class in each of the policies implemented here.  The advantage is that you'd avoid duplicating all of the code that's common to both policies.

Reply:
Refactored TimestampedAverage and NodeLatencyTracker into a separate class. But I don't think `onNodeSuccess` and `onNodeError` can be refactored to be a separate class, because they have to use `latencies`, a map of `NodeLatencyTracker`. Also, the LBPs have to extend `DefaultLoadBalancingPolicy`, which implements `RequestTracker`. So the two methods have to be implemented anyway.